### PR TITLE
More stacking contexts

### PR DESCRIPTION
**Pull request by staktrace**
For APZ and scrollframes and stuff

### CODE REVIEW

File: gfx/layers/wr/WebrenderCanvasLayer.cpp [hunk 1]
```diff
@@ -81,6 +81,8 @@ WebRenderCanvasLayer::RenderLayer(wrstate* aWRState)
     return;
   }
 
+  PushStackingContextsForScrollframes(aWRState);
```
Comment (kvark) [inline gfx/layers/wr/WebrenderCanvasLayer.cpp]: perhaps, let's do it RAII style?

File: gfx/layers/wr/WebrenderLayerManager.cpp [hunk 2]
```diff
@@ -62,6 +80,48 @@ WebRenderLayer::TransformedVisibleBoundsRelativeToParent()
...
+    Rect parentBounds = ParentStackingContextBounds(i);
+    bounds.MoveBy(-parentBounds.x, -parentBounds.y);
+    // Subtract the MT scroll position from the overflow here so that the WR
+    // scroll offset (which is the APZ async scroll component) always fits in
+    // the available overflow. If we didn't do this and WR did bounds checking
+    // on the scroll offset, we'd fail those checks.
+    overflow.MoveBy(bounds.x - scrollPos.x, bounds.y - scrollPos.y);
+    if (gfxPrefs::LayersDump()) {
+      printf_stderr("Popping stacking context id %" PRIu64 " with bounds=%s overflow=%s\n",
+      fm.GetScrollId(), Stringify(bounds).c_str(), Stringify(overflow).c_str());
```
Comment (kvark) [inline gfx/layers/wr/WebrenderLayerManager.cpp]: wrong indentation?

File: gfx/layers/wr/WebrenderLayerManager.cpp [hunk 3]
```diff
@@ -165,12 +226,34 @@ WebRenderLayerManager::EndTransaction(DrawPaintedLayerCallback aCallback,
 }
 
 void
+WebRenderLayerManager::ApplyAPZOffsets()
+{
+  // Note: this will probably set the same scroll offset multiple times
+  // because multiple layers will have the same scrollIds. We should just keep
```
Comment (kvark) [inline gfx/layers/wr/WebrenderLayerManager.cpp]: nit: rephrase with TODO

**Review by kvark — COMMENTED** [review]
I can't confirm that the actual APZ logic is correct, and I'd prefer RAII instead of Push/Pop, but otherwise looks healthy.

**Comment by staktrace** [comment]
Good points, I'll update the patches

**Comment by jrmuizel** [comment]
It turns out "Make async scrolling mostly work." was the commit that causes quarter rendering on macOS.

**Pull request merged**